[SelectionDAG] Folding  ZERO-EXTEND/SIGN_EXTEND poison to Poison value in getNode

diggerlin opened the pull request:
The PR will fix the issue https://github.com/llvm/llvm-project/issues/122728

This patch addresses the signed/zero extension of poison by using a poison value of the extended type instead of a constant zero of the extended type.

hubert-reinterpretcast commented:
Unfortunately, I think the IR semantics do not say that we can do this.

For example: The result of sign-extending `undef` must leave the left (more significant) bits either all ones or all zeroes. Using `undef` as the result of the sign-extension does not retain that property. Another way to see it is that "valid" realizations of the wider `undef` may have values that are impossible as the result of extending from a narrower type.

topperc commented:
I think we would be able to do this for poison, but not undef. Unfortunately SelectionDAG turns poison into undef.

diggerlin commented:
> The issue that the effect of `DeadArgumentEliminationPass` is not fully realized is shared with other platforms that require width extension of argument values such as RISC-V and LoongArch: https://godbolt.org/z/x55oqhT8h.

thanks, I will take a look the pass

topperc commented:
> > I think we would be able to do this for poison, but not undef. Unfortunately SelectionDAG turns poison into undef.
> 
> The IR documentation does not appear to support this. It says
> 
> > It is correct to replace a poison value with an [undef value](https://llvm.org/docs/LangRef.html#undefvalues) or any value of the type.
> 
> and
> 
> > This means that immediate undefined behavior occurs if a poison value is used as an instruction operand that has any values that trigger undefined behavior.
> 
> Any replacement of poison of the narrower type with a value (or `undef`) of the narrower type will still leave us with the same situation, and neither the call (which does not have `noundef` on the parameter of interest) nor the `sext`/`zext` conversions have immediate undefined behaviour tied to the `poison` argument of interest.
> 
> I think the direction to take here is to see if the middle-end can change the function parameter to be `noext`.

My understanding is that (zext/sext undef) needs to produce a value where all the extended bits are 0 for zext or match the original sign bit for sext. According to InstCombine, (sext/zext poison) can constant fold to poison in the larger type. If SelectionDAG preserved the difference between poison and undef, it could also constant fold to poison.

hubert-reinterpretcast commented:
> According to InstCombine, (sext/zext poison) can constant fold to poison in the larger type.

It sounds like you have more context than I do then. I believe that the description of `sext`, `zext`, etc. in the IR will benefit from this clarification. Are you open to posting a PR?

topperc commented:
> > According to InstCombine, (sext/zext poison) can constant fold to poison in the larger type.
> 
> It sounds like you have more context than I do then. I believe that the description of `sext`, `zext`, etc. in the IR will benefit from this clarification. Are you open to posting a PR?

There's a general statement in the Poison Values section. https://llvm.org/docs/LangRef.html#poison-values

```
Most instructions return ‘poison’ when one of their arguments is ‘poison’. A notable exception is the [select instruction](https://llvm.org/docs/LangRef.html#i-select). Propagation of poison can be stopped with the [freeze instruction](https://llvm.org/docs/LangRef.html#i-freeze).
```

diggerlin commented:
> The issue that the effect of `DeadArgumentEliminationPass` is not fully realized is shared with other platforms that require width extension of argument values such as RISC-V and LoongArch: https://godbolt.org/z/x55oqhT8h.

I added some printf into the `DeadArgumentElimination.cpp` , it run for AIX OS,  it is IR to IR pass which is invoked by `clang`(not invoked by `llc`).

in the line https://github.com/intel/llvm/blob/sycl/llvm/lib/Transforms/IPO/DeadArgumentElimination.cpp#L336, it converts the unused argument to `poison`.

the patch https://reviews.llvm.org/D125983 change the unused argument from `undef` to `poison`

 but according to the https://releases.llvm.org/11.0.0/docs/LangRef.html#poison-values
 
 there is description:
 
 There is currently no way of representing a poison value in the IR; they only exist when produced by operations such as [add](https://releases.llvm.org/11.0.0/docs/LangRef.html#i-add) with the nsw flag.
 
 in the patch https://reviews.llvm.org/D125983, the poison can be represented a argument.  
 
 do the document need to be modified?

hubert-reinterpretcast commented:
> but according to the https://releases.llvm.org/11.0.0/docs/LangRef.html#poison-values

The "up-to-date" version of the document is here: https://llvm.org/docs/LangRef.html#poison-values.
It indicates that `poison` can be used like a constant.

hubert-reinterpretcast commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -977,6 +977,18 @@ void SelectionDAGLegalize::LegalizeOp(SDNode *Node) {
   TargetLowering::LegalizeAction Action = TargetLowering::Legal;
   bool SimpleFinishLegalizing = true;
   switch (Node->getOpcode()) {
+  // FixMe: If the node represents a poison value, replace it with an undef
+  // value.
```
Can you explain what is to be fixed here? The text seems to explain what the code is doing (not what it should do differently).

hubert-reinterpretcast commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -9237,6 +9263,11 @@ SDValue SelectionDAG::getLoad(ISD::MemIndexedMode AM, ISD::LoadExtType ExtType,
   SDVTList VTs = Indexed ?
     getVTList(VT, Ptr.getValueType(), MVT::Other) : getVTList(VT, MVT::Other);
+
+  // FixedMe: lower poison to undef.
```
Same comment. Also: I think the convention is `FIXME`.

arsenm reviewed (commented):
> Title should be about introducing poison to the DAG, not something about arguments I don't understand

This PR as-is is not about introducing poison to the DAG (that's the how, not the what). The actual operative change is the handling of poison for the width extensions.

Would "Introduce poison to the DAG; address signed/zero extend of poison" work?

arsenm commented:
> This PR as-is is not about introducing poison to the DAG (that's the how, not the what). The actual operative change is the handling of poison for the width extensions.

These should not be mixed in. Introducing poison should be separate from any of this other stuff

diggerlin commented:
There are some formatters that failed in the check for code formatting: undef deprecator. I think this is related to my use of the SDValue SelectionDAG::getUNDEF(EVT VT) function in the patch.

```
// Lower poison to undef.
if (Ptr.getNode()->isPoison())
  Ptr = getUNDEF(Ptr.getValueType());
```

```
case ISD::POISON: {
  SDValue UndefNode = DAG.getUNDEF(Node->getValueType(0));
  ReplaceNode(Node, UndefNode.getNode());
  break;
}
```

If I can no longer use this function, the patch will become quite large. Can I still use getUNDEF in this patch while keeping the undef deprecator, and then introduce a follow-up patch to remove getUNDEF later? 

@hubert-reinterpretcast , @topperc

arsenm commented:
> > do you means we should have two separate patches? one for  introducing poison to the DAG, other patch for handling of poison for the width extensions?

Yes

arsenm commented:
> If I can no longer use this function, the patch will become quite large. Can I still use getUNDEF in this patch while keeping the undef deprecator, and then introduce a follow-up patch to remove getUNDEF later?
> 
> @hubert-reinterpretcast , @topperc

Ignore the undef deprecator, it is broken. It should not be complaining about getUNDEF (or really anything in codegen)

diggerlin commented:
I created a separate patch 
https://github.com/llvm/llvm-project/pull/125883

after the patch landed,  I will return back the patch.

amy-kwan commented on llvm/test/CodeGen/PowerPC/undef-args.ll:
```diff
@@ -0,0 +1,221 @@
+;; Tests that load 0 into argument registers for unused arguments are eliminated. 
+
+; RUN: llc -verify-machineinstrs -mtriple powerpc-ibm-aix-xcoff -debug-only=isel \
+; RUN:   %s -o - 2>&1 | FileCheck --check-prefix=CHECKSEL32 %s
```
Nit on the check name.
```suggestion
; RUN:   %s -o - 2>&1 | FileCheck --check-prefix=CHECKISEL32 %s
```

amy-kwan commented on llvm/test/CodeGen/PowerPC/undef-args.ll:
```diff
@@ -0,0 +1,221 @@
+; RUN:   %s -o - 2>&1 | FileCheck --check-prefix=CHECKSEL32 %s
+
+; RUN: llc -verify-machineinstrs -mtriple powerpc64-ibm-aix-xcoff -debug-only=isel \
+; RUN:   %s -o - 2>&1 | FileCheck --check-prefix=CHECKSEL64 %s
```
Nit on the check name.
```suggestion
; RUN:   %s -o - 2>&1 | FileCheck --check-prefix=CHECKISEL64 %s
```

amy-kwan commented on llvm/include/llvm/CodeGen/SelectionDAGNodes.h:
```diff
@@ -690,8 +690,17 @@ END_TWO_BYTE_PACK()
+  /// - When UndefOnly is true, returns true only for UNDEF.
+  /// - When UndefOnly is false, returns true for both UNDEF and POISON.
+  /// @param UndefOnly Determines whether to check only for UNDEF.
+  bool isUndef(bool UndefOnly = false) const {
```
Question: So currently, we just want this true for both undef and poison, right? Since I don't think we explicitly set UndefOnly to true anywhere.

amy-kwan reviewed (commented):
Since this patch is dependent on https://github.com/llvm/llvm-project/pull/125883, when the https://github.com/llvm/llvm-project/pull/125883 lands, this will need to be rebased/updated. I also added comments to the previous patch.

diggerlin committed:
lower zero(sign) extend poison to poison

diggerlin commented:
I rebased the patch based on the [[SelectionDAG] Introducing a new ISD::POISON SDNode to represent the poison value in the IR](https://github.com/llvm/llvm-project/pull/125883#top),I would appreciate any feedback on it.

diggerlin commented:
I saw that you recently committed some AMDGPU-related test cases. Since I changed SIGN_EXTEND(POISON) and ZERO_EXTEND(POISON) to just POISON in my patch, I’ve updated the test cases accordingly. Would you be willing to help review them?  @arsenm

diggerlin committed:
change AMDGPU test case

diggerlin committed:
minor change

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -6241,6 +6241,10 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
       return getNode(OpOpcode, DL, VT, N1.getOperand(0), Flags);
     }
+
+    if (N1.isPoison())
+      return getPOISON(VT);
```
Slightly annoying that this is referring to OpOpcode by another method

arsenm reviewed (commented):
Can you fix the title to state this is about folding poison inputs in getNode

diggerlin committed:
address comment

diggerlin commented:
Whenever you have a moment, I’d appreciate any further feedback you might have.

nico commented:
It looks like this broke check-llvm everywhere. I've reverted it for now.

qinkunbao commented:
Hi,

This PR broke the build bot.
https://lab.llvm.org/buildbot/#/builders/24/builds/7600
https://lab.llvm.org/buildbot/#/builders/94/builds/6413
https://lab.llvm.org/buildbot/#/builders/85/builds/7898

diggerlin commented:
I created a patch https://github.com/llvm/llvm-project/pull/136636 to fix the test cases which cause the buildbot fail
